allow users to set sample rate and add per iteration latency breakdowns

zhaojuanmao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #53148 clang format reducer and logger files
* #53162 fix data type logging
* #53149 add is_multi_device_module logging field
* **#53145 allow users to set sample rate and add per iteration latency breakdowns**

add a new API to allow users to set sample rate for runtime stats, also add per iteration latency breakdowns to DDPLoggingData struct. e.g.
if users set sample rate to be 1, they can analyze per iteration latency change over time (not avged)

Differential Revision: [D26763957](https://our.internmc.facebook.com/intern/diff/D26763957/)

zhaojuanmao committed:
allow users to set sample rate and add per iteration latency breakdowns

add a new API to allow users to set sample rate for runtime stats, also add per iteration latency breakdowns to DDPLoggingData struct. e.g.
if users set sample rate to be 1, they can analyze per iteration latency change over time (not avged)

Differential Revision: [D26763957](https://our.internmc.facebook.com/intern/diff/D26763957/)

[ghstack-poisoned]

SciPioneer commented on torch/lib/c10d/reducer.hpp:
```diff
@@ -379,6 +385,9 @@ class Reducer {
   void record_backward_comm_end_time();
 
+  int get_ddp_runtime_logging_sample_rate();
+  int ddp_runtime_logging_sample_rate_ = 0;
```
Why not directly use kDDPRuntimeLoggingSampleRate?

SciPioneer commented on torch/nn/parallel/distributed.py:
```diff
@@ -1277,3 +1277,27 @@ def _set_params_and_buffers_to_ignore_for_model(
+        "kDDPRuntimeLoggingSampleRate=100" training iterations.
+        """
+        if sample_rate <= 0:
+            raise ValueError("DDP runtime logging sample rate should be larger than 0")
```
nit: larger/greater

SciPioneer commented on torch/nn/parallel/distributed.py:
```diff
@@ -1277,3 +1277,27 @@ def _set_params_and_buffers_to_ignore_for_model(
+        """
+        return self.logger._get_ddp_logging_data()
+
+    def set_ddp_runtime_logging_sample_rate(self, sample_rate):
```
The term "sample rate" actually can be a bit misleading. Normally sampling rate is a fractional number between 0 and 1. I think "sampling period" or "sampling interval" can  be a better term here.

zhaojuanmao committed:
Update on "allow users to set sample rate and add per iteration latency breakdowns"


add a new API to allow users to set sample rate for runtime stats, also add per iteration latency breakdowns to DDPLoggingData struct. e.g.
if users set sample rate to be 1, they can analyze per iteration latency change over time (not avged)

Differential Revision: [D26763957](https://our.internmc.facebook.com/intern/diff/D26763957/)

[ghstack-poisoned]

rohan-varma commented on torch/nn/parallel/distributed.py:
```diff
@@ -1305,3 +1305,27 @@ def _set_params_and_buffers_to_ignore_for_model(
+        after 10 iterations runtime stats are recorded once every
+        "kDDPRuntimeLoggingSampleRate=100" training iterations.
+        """
+        if sample_rate <= 0:
```
Should we check that the arg is an int and greater/equal to 1? Something like 0.5 for example doesn't make sense (and wouldn't that be cast down to 0 if it's passed into cpp as an int?)

zhaojuanmao committed:
Update on "allow users to set sample rate and add per iteration latency breakdowns"


add a new API to allow users to set sample rate for runtime stats, also add per iteration latency breakdowns to DDPLoggingData struct. e.g.
if users set sample rate to be 1, they can analyze per iteration latency change over time (not avged)

Differential Revision: [D26763957](https://our.internmc.facebook.com/intern/diff/D26763957/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "allow users to set sample rate and add per iteration latency breakdowns"


add a new API to allow users to set sample rate for runtime stats, also add per iteration latency breakdowns to DDPLoggingData struct. e.g.
if users set sample rate to be 1, they can analyze per iteration latency change over time (not avged)

Differential Revision: [D26763957](https://our.internmc.facebook.com/intern/diff/D26763957/)

[ghstack-poisoned]
